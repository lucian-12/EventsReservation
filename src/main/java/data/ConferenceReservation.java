package data;

public class ConferenceReservation extends Reservation {
    public ConferenceReservation(String reservationID, int nbSeats, Event event) {
        super(reservationID, nbSeats, event);
    }

    @Override
    public int calculateAmount() {
        int result = 40000;
        if (getNbSeats() > 3) {
            result += 1000 * (getNbSeats() - 3);
        }
        return result;
    }

    @Override
    int calculateVolumeCredits() {
        int result = 0;
        result += Math.max(getNbSeats() - 2, 0);
        return result;
    }
}
