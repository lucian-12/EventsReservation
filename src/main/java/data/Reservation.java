package data;

public abstract class Reservation {

    private String reservationID;
    private int nbSeats;
    private Event event;

    public Reservation(String reservationID, int nbSeats, Event event) {
        this.reservationID = reservationID;
        this.nbSeats = nbSeats;
        this.event = event;
    }

    public Event getEvent() {
        return event;
    }

    public Reservation setEvent(Event event) {
        this.event = event;
        return this;
    }

    public String getReservationID() {
        return reservationID;
    }

    public Reservation setReservationID(String reservationID) {
        this.reservationID = reservationID;
        return this;
    }

    public int getNbSeats() {
        return nbSeats;
    }

    public Reservation setNbSeats(int nbSeats) {
        this.nbSeats = nbSeats;
        return this;
    }

     public abstract int calculateAmount();

    int calculateVolumeCredits() {
        int result = 0;
        // add volume credits
        result += Math.max(getNbSeats() - 2, 0);
        // add extra credit for every workshop attendee
        if ("workshop".equals(getEvent().getType()))
            result += Math.floor(getNbSeats());
        return result;
    }
}
