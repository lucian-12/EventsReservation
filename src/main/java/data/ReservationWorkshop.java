package data;

public class ReservationWorkshop extends Reservation {
    public ReservationWorkshop(String reservationID, int nbSeats, Event event) {
        super(reservationID, nbSeats, event);
    }

    @Override
    public int calculateAmount() {
        int result = 30000;
        if (getNbSeats() > 2) {
            result += 10000 + 500 * (getNbSeats() - 2);
        }
        result += 300 * getNbSeats();
        return result;
    }

    @Override
    int calculateVolumeCredits() {
        int result = 0;
        result += Math.max(getNbSeats() - 2, 0);
        result += Math.floor(getNbSeats());
        return result;
    }
}
