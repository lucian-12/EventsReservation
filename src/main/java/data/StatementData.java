package data;

import java.util.List;

public class StatementData {
    private String customer;
    private List<Reservation> reservations;
    private int totalAmount;
    private int totalVolumeCredits;

    public void setCustomer(String customer) {

        this.customer = customer;
    }

    public String getCustomer() {
        return customer;
    }

    public void setReservations(List<Reservation> reservations) {
        this.reservations = reservations;
    }

    public List<Reservation> getReservations() {
        return reservations;
    }

    int calculateTotalVolumeCredits() {
        return getReservations().stream().mapToInt(Reservation::calculateVolumeCredits).sum();
    }

    int calculateTotalAmount() {
        return getReservations().stream().mapToInt(Reservation::calculateAmount).sum();
    }

    public void setTotalAmount(int totalAmount) {
        this.totalAmount = totalAmount;
    }

    public int getTotalAmount() {
        return totalAmount;
    }

    public void setTotalVolumeCredits(int totalVolumeCredits) {
        this.totalVolumeCredits = totalVolumeCredits;
    }

    public int getTotalVolumeCredits() {
        return totalVolumeCredits;
    }

}