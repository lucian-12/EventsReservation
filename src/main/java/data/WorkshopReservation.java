package data;

public class WorkshopReservation extends Reservation {
    public WorkshopReservation(String reservationID, int nbSeats, Event event) {
        super(reservationID, nbSeats, event);
    }

    @Override
    public int calculateAmount() {
        int result = 30000;
        if (getNbSeats() > 2) {
            result += 10000 + 500 * (getNbSeats() - 2);
        }
        result += 300 * getNbSeats();
        return result;
    }
}
