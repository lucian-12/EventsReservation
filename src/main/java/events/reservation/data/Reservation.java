package events.reservation.data;

public class Reservation {

    private String reservationID;
    private int nbSeats;
    private Event event;

    public Reservation(String reservationID, int nbSeats, Event event) {
        this.reservationID = reservationID;
        this.nbSeats = nbSeats;
        this.event = event;
    }

    public Event getEvent() {
        return event;
    }

    public Reservation setEvent(Event event) {
        this.event = event;
        return this;
    }

    public String getReservationID() {
        return reservationID;
    }

    public Reservation setReservationID(String reservationID) {
        this.reservationID = reservationID;
        return this;
    }

    public int getNbSeats() {
        return nbSeats;
    }

    public Reservation setNbSeats(int nbSeats) {
        this.nbSeats = nbSeats;
        return this;
    }

    public int calculateAmount() {
        int result;

        switch (getEvent().getType()) {
            case "conference":
                result = 40000;
                if (getNbSeats() > 3) {
                    result += 1000 * (getNbSeats() - 3);
                }
                break;
            case "workshop":
                result = 30000;
                if (getNbSeats() > 2) {
                    result += 10000 + 500 * (getNbSeats() - 2);
                }
                result += 300 * getNbSeats();
                break;
            default:
                throw new Error("unknown type: ${event.type}");
        }
        return result;
    }

    public int calculateTotalVolumeCredits() {
        int result = Math.max(getNbSeats() - 2, 0);
        if ("workshop".equals(getEvent().getType())) {
            result += Math.floor(getNbSeats());
        }
        return result;
    }
}
