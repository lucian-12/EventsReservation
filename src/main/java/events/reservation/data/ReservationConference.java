package events.reservation.data;

public class ReservationConference extends Reservation {
    public ReservationConference(String reservationID, int nbSeats, Event event) {
        super(reservationID, nbSeats, event);
    }

    @Override
    public int calculateAmount() {
        int result = 40000;
        if (getNbSeats() > 3) {
            result += 1000 * (getNbSeats() - 3);
        }
        return result;
    }

    @Override
    public int calculateTotalVolumeCredits() {
        return Math.max(getNbSeats() - 2, 0);
    }
}
