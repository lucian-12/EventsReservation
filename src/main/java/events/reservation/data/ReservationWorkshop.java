package events.reservation.data;

public class ReservationWorkshop extends Reservation {
    public ReservationWorkshop(String reservationID, int nbSeats, Event event) {
        super(reservationID, nbSeats, event);
    }

    @Override
    public int calculateAmount() {
        int result = 30000;
        if (getNbSeats() > 2) {
            result += 10000 + 500 * (getNbSeats() - 2);
        }
        result += 300 * getNbSeats();
        return result;
    }

    @Override
    public int calculateTotalVolumeCredits() {
        int result = Math.max(getNbSeats() - 2, 0);
        result += Math.floor(getNbSeats()) / 3;
        return result;
    }

}
