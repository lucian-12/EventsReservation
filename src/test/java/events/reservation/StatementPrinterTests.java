package events.reservation;

import events.reservation.data.*;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.approvaltests.Approvals.verify;

public class StatementPrinterTests {

    @Test
    void exampleStatement() {

        Invoice invoice = new Invoice("BigCo", List.of(
                new ReservationConference("dev", 5, new Event("DevConference", "conference")),
                new ReservationWorkshop("web", 3, new Event("WebWorkshop", "workshop")),
                new ReservationConference("ml", 4, new Event("MachineLearning", "conference"))));

        StatementPrinter statementPrinter = new StatementPrinter();
        String result = statementPrinter.print(invoice);

        verify(result);
    }
}
