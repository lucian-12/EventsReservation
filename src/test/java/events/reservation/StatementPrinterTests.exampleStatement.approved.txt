Statement for BigCo
  DevConference: $420.00 (5 seats)
  WebWorkshop: $414.00 (3 seats)
  MachineLearning: $410.00 (4 seats)
Amount owed is $1,244.00
You earned 7 credits
